### ISSUE TITLE
descriptive errors

### ISSUE THREAD

**Issue by timruffles** [philbooth/check-types.js#44]
love the library. would be great to get descriptive errors out, something like:

```javascript
var gitSha = /a/.test.bind(/^[a-f0-8]{40}$/);
gitSha.message = "'%v' is not a valid git sha";

var err = check.map.describe({
  shaA: "c19d86ba2dfb49986a0db052d394ded29da71775",
  shaB: "zzz0",
}, {
  shaA: gitSha,
  shaB: gitSha,
});

err.message === `type invalid: { shaB: "'zzz0' is not a valid git sha" }` // true
```

If you're interested in the approach, I'd be happy to PR.

**Comment by philbooth** [comment]
I definitely like the use case.

Couple of things I'm undecided on:

1. Making `message

**Comment by timruffles** [comment]
Seems like it would make it a lot more typing/verbose tho?

Re modifying, maybe making a method that returns a derived predicate with a new message instead (have .message of the builtins as non-`writable/configurable`)?

```javascript
var gitSha = /a/.test.bind(/^[a-f0-8]{40}$/);
gitSha.message = "'%v' is not a valid git sha";

var validUserId = check.integer.derive("'%s' is not a valid userId");

var err = check.map.describe({
  shaA: "c19d86ba2dfb49986a0db052d394ded29da71775",
  shaB: "zzz0",
}, {
  shaA: gitSha,
  shaB: gitSha,
  userId: validUserId,
});
```

**Comment by philbooth** [comment]
Yep, fair point about the verbosity.

Really like your extra method idea and am rolling a few alternative names round my head. What do you think of the following?

* `annotate`
* `comment`
* `message`
* `reason`

**Comment by philbooth** [comment]
Those were alternatives for `derive` btw, not `describe` (which I like).

**Comment by timruffles** [comment]
All good words - my thinking for `derive` was to suggest you're taking a general type like `int` and giving it an application specific meaning.

**Comment by philbooth** [comment]
So the reason I dislike `derive` a bit is because it suggests a child building on the inherited logic from its parent, e.g. `positiveInteger` as a specialisation of `integer`. But really the actual deductive logic of the predicate isn't changing, we're just saying "instead of `false`, return a message". Hence thinking along the lines of `annotate` etc.

But, I realise I'm bike-shedding to the extreme here. What I actually meant to say was: "Thanks Tim, awesome suggestion, I'll definitely accept a PR if you have time to raise one."

:)

**Comment by timruffles** [comment]
That makes complete sense - what about 'alias' or 'synonym'? Second is used
in Haskell

**Comment by philbooth** [comment]
Yep, `alias` > `synonym` imho, but both are good and I'm happy with either.

**Comment by MareoRaft** [comment]
Excuse me if I am posting this in the wrong place.  I am getting the message "`Uncaught Error: Invalid object`".  I wish it would tell me which `check.assert` in my code caused the failure (line number), as well as the actual type and value of the variable it was checking.  I don't know what is and is not possible.

**Comment by philbooth** [comment]
@MareoRaft, this probably is the wrong place so I've opened a new issue to answer your question: #51

**Comment by kress95** [comment]
Just a function returning something like {valid: true, received: "typename", expected: "typename"} would be enough for me :+1:

**Comment by milesj** [comment]
Any progress on this?
